pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+        properties["Total Qubits"] = get_qubit_count(program)
+        properties["Compile Time"] = time.time() / 1000.0 - start_time
+        results.append(properties)
+        
+    results = aos_to_soa(results)
+    
+    # Filters results if specific ones are requested
+    if to_include != None:
+        filtered = {}
+        for val in to_include:
+            if val in results:
+                filtered[val] = results[val]
+        results = filtered
+    
+    # Removes properties that are always zero, if requested
+    if remove_zeros:
+        for key in results.keys():
+            if sum(results[key]) == 0:
+                del results[key]
+    
+    return (input_list, results)
+
+def plot_analysis(x_points, analysis, title, x_axis, y_axis, bounds=None):
+    '''
+    Takes results from the analyze function and plots them using plotly
+    :param analysis: The analysis gathered by the analyze function
+    :param title: The title to display on the plotted graph
+    :param x_axis: A label for the x axis
+    :param y_axis: A label for the y axis
+    :param bounds: An optional dictionary of names to functions, which will
+                   be plotted against the given x points. This is useful
+                   for comparing your gate growth to the requested growth
+    :param x_points: A list of x points to use for your data (i.e. input
+                     for your programs, "n")
+    '''
+    runtime = {}
+    data = []
+    
+    r = lambda: random.randint(0,255)
+    
+    for val in analysis.keys():   # For each property
+        
+        # Create the graph
+        this_data = analysis[val]
+        trace = go.Scatter(
+            x = x_points,
+            y = this_data,
+            name = val,
+            line = dict(
+                color = ('rgb(%d, %d, %d)' % (r(), r(), r())),
+                width = 4)
+        )
+        data.append(trace)
+        
+        # Find the polynomial runtime
+        runtime[val] = get_runtime_approximate(x_points, this_data)
+        #print str(val) + ": " + str(runtime[val][0]) + "n^" + str(runtime[val][1])
+        
+    if bounds is not None:
+        for name in bounds.keys():
+            # Create the graph
+            function = bounds[name]
+            this_data = [function(i) for i in x_points]
+            trace = go.Scatter(
+                x = x_points,
+                y = this_data,
+                name = name,
+                line = dict(
+                    color = ('rgb(%d, %d, %d)' % (r(), r(), r())),
+                    width = 4,
+                    dash = 'dash')
+            )
+            data.append(trace)
+        
+
+    # Edit the layout
+    layout = dict(title = title,
+                  xaxis = dict(title = x_axis),
+                  yaxis = dict(title = y_axis),
+                  )
+
+    fig = dict(data=data, layout=layout)
+    plotly.offline.plot(fig)
+
+def aos_to_soa(aos):
```

Comment:
This method is specific to my implementation in that I know the first structure will have all of the keys that I need, but this may not be true in a general sense if someone else where to use it for their own code. Should we improve it and maybe put it into a util class?